cabal-install-solver/src/Distribution/Solver/Types/ResolverPackage.hs

```diff
@@ -23,30 +25,36 @@ import Distribution.Simple.Utils (ordNub)
 --
 -- This is like the 'InstallPlan.PlanPackage' but with fewer cases.
 --
-data ResolverPackage loc = PreExisting InstSolverPackage
-                         | Configured  (SolverPackage loc)
+-- The constraint scope field refers to the scope in which this package was resolved.
+data ResolverPackage loc = PreExisting InstSolverPackage ConstraintScope
+                         | Configured  (SolverPackage loc) ConstraintScope
   deriving (Eq, Show, Generic)
 
 instance Binary loc => Binary (ResolverPackage loc)
 instance Structured loc => Structured (ResolverPackage loc)
 
 instance Package (ResolverPackage loc) where
-  packageId (PreExisting ipkg)     = packageId ipkg
-  packageId (Configured  spkg)     = packageId spkg
+  packageId (PreExisting ipkg _)     = packageId ipkg
+  packageId (Configured  spkg _)     = packageId spkg
 
 resolverPackageLibDeps :: ResolverPackage loc -> CD.ComponentDeps [SolverId]
-resolverPackageLibDeps (PreExisting ipkg) = instSolverPkgLibDeps ipkg
-resolverPackageLibDeps (Configured spkg) = solverPkgLibDeps spkg
+resolverPackageLibDeps (PreExisting ipkg _) = instSolverPkgLibDeps ipkg
+resolverPackageLibDeps (Configured spkg _) = solverPkgLibDeps spkg
 
 resolverPackageExeDeps :: ResolverPackage loc -> CD.ComponentDeps [SolverId]
-resolverPackageExeDeps (PreExisting ipkg) = instSolverPkgExeDeps ipkg
-resolverPackageExeDeps (Configured spkg) = solverPkgExeDeps spkg
+resolverPackageExeDeps (PreExisting ipkg _) = instSolverPkgExeDeps ipkg
+resolverPackageExeDeps (Configured spkg _) = solverPkgExeDeps spkg
+
+resolverPackageConstraintScope :: ResolverPackage oc -> ConstraintScope
+resolverPackageConstraintScope (PreExisting _ cs) = cs
+resolverPackageConstraintScope (Configured _ cs) = cs
 
 instance IsNode (ResolverPackage loc) where
   type Key (ResolverPackage loc) = SolverId
-  nodeKey (PreExisting ipkg) = PreExistingId (packageId ipkg) (installedUnitId ipkg)
-  nodeKey (Configured spkg) = PlannedId (packageId spkg)
+  nodeKey (PreExisting ipkg _) = PreExistingId (packageId ipkg) (installedUnitId ipkg)
+  nodeKey (Configured spkg _) = PlannedId (packageId spkg)
```

Comment:
The problem is that when we construct a graph of resolver packages in the solver the keys don't have this information (even though the nodes do); thus same packages resolved in different constraint scopes will be collapsed and only one copy will survive.

In cabal-install, when we construct a GenericInstallPlan from the graph of resolver packages we have only kept one of each package, even if originally many could have existed in different scopes.